### PR TITLE
MITM: recreating private/public keys all the time

### PR DESCRIPTION
**Pull request by stiray**
... has huge impact on speed of processing, with this change the RSA and ECDSA keys are created within "init()" and used when needed. This is quick fix, please set it to some configuration variable if you consider that it is useful to generate them on each request. I dont think this would impact the security on client side.

### CODE REVIEW

File: signer.go [hunk 1]
```diff
@@ -36,8 +40,86 @@ func hashSortedBigInt(lst []string) *big.Int {
...
+	key := strings.Join(hosts[:], "/")
+	value, exist := crtCache.Get(key)
+	if exist {
+		cert = value.(*tls.Certificate)
+		return
+	}
+
+    cert, err = generateCertificate(ca, hosts)
+
+	crtCache.Add(key, cert, cache.DefaultExpiration)
```
Comment (engineering-this) [inline signer.go]: Race condition scenario:

Two concurrent calls to signHost for hostname with not yet cached crt.
The second call will fail to add cache, as it is existing, although it didn't exist when line #50 was called.

Comment (hazcod) [reply signer.go:58]: @engineering-this: how would you solve that race condition?

Comment (engineering-this) [reply signer.go:58]: I'd go with a semaphore.

```go
var semaphore = make(chan struct{}, 1)

func signHost(ca tls.Certificate, hosts []string) (cert *tls.Certificate, err error) {
	semaphore <- struct{}{}
	defer func() {
		<-semaphore
	}()
	//...
```

**Comment by hazcod** [comment]
@stiray @elazarl @engineering-this : see https://github.com/hazcod/goproxy/commit/152f8657271878d3f574290108e33e2d6b70c348 for a possibly better implementation.

**Comment by stiray** [comment]
Ok, I am not using github and I dont even intend to as I have my own git server and I am quite happy with command line so I wont be clicking anything here any more... 

Regarding the "race condition" which was probably merged due to clicking to "resolve whatever" which was a mistake if the fix came into code.

The race condition was put there deliberately to avoid synchronizing multiple threads using synchronization object at a cost that here and then the certificate is generated without any reason which is performance wise better decision.

But nvm me, this was a drive by code, it have my own repo/fork so please do whatever feels appropriate for you.

**Comment by ErikPelli** [comment]
Proxy server now has a `CertStore` field that caches the certificates and fix the issue that you explained. This merge request is not needed anymore.

**Pull request closed without merging**